# Fundamentals of Python
## Hello World!
Program. It is the tool to provide the command to computers. We enter what we want computers to do, and computers executes the calculation.  
 Modern computers can do many things. Not only can they calculate some formula, but they can also draw graphs, process image, and learn somethig with neural network. This means that engineers should know how to tell computers to do so.  
 In the early days, learning programming is the process of memorizing the syntax, and understand how to make use of them. But now, as there are too many commands to use, we cannot memorize all things. What the modern engineers should do is not memorizing, but understand how to google it.  
 <br>
 First, let's do a "customary event" of start learning programming. Enter the program below.  
```python
print("Hello World")
```
 Execute this command by clicking the button "▷" or type "Shift + Enter".


In [1]:
print("Hello World")

Hello World


 Good. Now you can make your computer say "Hello World."  
 What you write now is actually very important. After some calculation, we want computers to show us the result. The simplest way to show the result is the "standard output", which you have just done.  
 In the next chapter, you learn how to write simple calculation like four arithmetic operations, and each time, you will check whether the calclation succeed or not using the command ```print```.

## Operators in Python
 In this section, you will learn operators. Operators are sets of symbols denoting an operation(e.g. +, -, <, >). 
 First, enter the program below.  

```python
print(7+3)
print(7-3)
print(7*3)
print(7/3)
print(7%3)
print(7//3)

print(7**3)
```

In [2]:
print(7+3)
print(7-3)
print(7*3)
print(7/3)
print(7//3)
print(7%3)

10
4
21
2.3333333333333335
2
1


In [3]:
print(7**3)

343


 In python, ```+``` represents addition, ```-``` represents subtraction, ```*``` represents multiplication, and ```/``` represents division. Note that the result of ```/``` is a float value.(float value≒Decimal) If you want to calculate quotient and remainder, use ```//``` and ```%``` respectively.  
 When you calculate the power, the operator to use is ```**```. (```a**b``` means $a^b$)  
 Next, run the following program.

```python
print(7==3)
print(3==3)
print(7!=3)
print(3!=3)

print(7>3)
print(3>3)
print(7>=3)
print(3>=3)

print(7<3)
print(3<3)
print(7<=3)
print(3<=3)
```

In [4]:
print(7==3)
print(3==3)
print(7!=3)
print(3!=3)

False
True
True
False


In [5]:
print(7>3)
print(3>3)
print(7>=3)
print(3>=3)

True
False
True
True


In [6]:
print(7<3)
print(3<3)
print(7<=3)
print(3<=3)

False
False
False
True


 The output will be "True" or "False". Needless to say, "True" means that formula is correct, and "False" means that the formula is not correct. We call True/False value "Boolean".  
 Note that the operator to assess whether the value is same or not is ```==```, not "=". Also, remember that ```!=``` stands for "≠".  
<br>
 Next, run the following code.
 
```python
print(3<7 and 7<10)
print(3<7 and 7<5)
print(3<0 and 7<5)

print(3<7 or 7<10)
print(3<7 or 7<5)
print(3<0 or 7<5)
```

In [7]:
print(3<5 and 3<7)
print(3<0 and 3<7)
print(3<0 and 3<1)

True
False
False


In [8]:
print(3<7 or 7<10)
print(3<7 or 7<5)
print(3<0 or 7<5)

True
True
False


 The keyword ```and```, and ```or``` assess two(or more) boolean value. ```and``` returns True only when both of two values are True. ```or``` returns True when at least one is True.  
```python
(True and True) == True
(True and False) == False
(False and True) == False
(False and False) == False
(True or True) == True
(True or False) == True
(False or True) == True
(False or False) == False
```

 Next, run this code
```python
print(1+2*3)
print((1+2)*3)
```

In [9]:
print(1+2*3)
print((1+2)*3)

7
9


 You can see that the multiplication has the priority over addition. When you want to do addition first, then you must enclose with parentheses.  
 The priority of the operators is defined like this.  

|priority|meaning|signature|
|:-:|:-:|:-:|
|1|formula inside ()|(......)|
|2|power|\*\*|
|3|sign|+X, -X|
|4|multiplecation, division|\*, /, %|
|5|addition, subtraction|+, -|
|6|comparison|<, <=, >, >=, !=, ==|
|7|and, or|and, or|

## Number and String
When you write programs, you must know about types.  
......-2,-1,0,1,2,.... are called **"intergers"** or **"int"**.<br>
1.23 or 1.23E10 are called **"float"**, because the decimal point goes up and down as if it was a boat that floats on the sea.  
When programming, the word **"string"** or **"str"** is the sequence of characters. String values act differently compared to numbers.  
Here are the example. Run the following code.
```python
print(11+12)
print("11"+"12")
```

In [10]:
print(11+12)
print("11"+"12")

23
1112


```11+12``` is 23, but ```"11"+"12"``` becomes 1112! This is the difference between number and string.  
When you enclose something inside quotes, that value is taken as a string. ```11``` is a number, but ```"11"``` is a string and computer take this as a two characters '1' and '1'.  
The operator ```+``` is a addition **only when** you place numbers before and after. If you place strings around operator ```+```, then it will be taken as a **concatenation.**  
```python
#If you run the following code, the error will occur.
#Do not try to add string and numbers!
print(11+"12")
print("11"+12)
```

In [11]:
print(11+"12")

TypeError: unsupported operand type(s) for +: 'int' and 'str'

As you see, the addition between int(integer) and str(string) is not allowed.

In [12]:
print("11"+12)

TypeError: can only concatenate str (not "int") to str

Also, the addition between str(string) and int(integer) is not allowed.

Next, run the following code.
```python
print("12"*2)
```

In [13]:
print("12"*2)

1212


Multiplication between string and integer is allowed, and it will concatenate words two (or more) times.

## Type Checking
Run the following code.
```python
print(type(12))
print(type(1.2))
print(type("12"))
```

In [14]:
print(type(12))
print(type(1.2))
print(type("12"))

<class 'int'>
<class 'float'>
<class 'str'>


As you can see, the function ```type()``` will tell us what the type of value is.

## Type Casting
You may wondering how to chage the type of values. You can do this by using ```int()```, ```float()```, and ```str()```  
Run the following code.
```python
print("123")
print(type("123"))
print(int("123"))
print(type(int("123")))
```

In [15]:
print("123")
print(type("123"))
print(int("123"))
print(type(int("123")))

123
<class 'str'>
123
<class 'int'>


Next run the following code.
```python
print(123)
print(type(123))
print(str(123))
print(type(str(123)))
```

In [16]:
print(123)
print(type(123))
print(str(123))
print(type(str(123)))

123
<class 'int'>
123
<class 'str'>


The function ```int()``` can convert string into integer, and the function ```str()``` can convert integer into string.  
Here is an example of the usage of type casting. Run the following code.
```python
print("3+4=" + str(3+4))
```

In [17]:
print("3+4=" + str(3+4))

3+4=7


The program tries to calculate the concatenation of ```"3+4="``` and ```str(3+4)```, both of which are strings.  
Next, the program tries to cast ```3+4``` to string, which means it changes 7(integer) into "7"(string).  
As a result, the outcome of the program becomes the string of "3+4=7".  
<br>
You can also cast between string to float.  
Run the following code.
```python
print(str(1.2))
print(type(str(1.2)))
print(float("1.2"))
print(type(float("1.2")))

print("1.2+3.4=" + str(1.2+3.4))
```

In [18]:
print(str(1.2))
print(type(str(1.2)))
print(float("1.2"))
print(type(float("1.2")))

1.2
<class 'str'>
1.2
<class 'float'>


In [19]:
print("1.2+3.4=" + str(1.2+3.4))

1.2+3.4=4.6
